service/listeners.go

```diff
@@ -0,0 +1,333 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"sync/atomic"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+)
+
+// The implementations of listeners for different network types are not
+// interchangeable. The type of listener depends on the network type.
+type Listener = io.Closer
+
+// StreamListener is a network listener for stream-oriented protocols that
+// accepts [transport.StreamConn] connections.
+type StreamListener interface {
+	// Accept waits for and returns the next connection to the listener.
+	AcceptStream() (transport.StreamConn, error)
+
+	// Close closes the listener.
+	// Any blocked Accept operations will be unblocked and return errors. This
+	// stops the current listener from accepting new connections without closing
+	// the underlying socket. Only when the last user of the underlying socket
+	// closes it, do we actually close it.
+	Close() error
+
+	// Addr returns the listener's network address.
+	Addr() net.Addr
+}
+
+type TCPListener struct {
+	ln *net.TCPListener
+}
+
+var _ StreamListener = (*TCPListener)(nil)
+
+func (t *TCPListener) AcceptStream() (transport.StreamConn, error) {
+	return t.ln.AcceptTCP()
+}
+
+func (t *TCPListener) Close() error {
+	return t.ln.Close()
+}
+
+func (t *TCPListener) Addr() net.Addr {
+	return t.ln.Addr()
+}
+
+type acceptResponse struct {
+	conn transport.StreamConn
+	err  error
+}
+
+type OnCloseFunc func() error
+
+type virtualStreamListener struct {
+	listener    StreamListener
+	acceptCh    chan acceptResponse
+	closeCh     chan struct{}
+	onCloseFunc OnCloseFunc
+}
+
+var _ StreamListener = (*virtualStreamListener)(nil)
+
+func (sl *virtualStreamListener) AcceptStream() (transport.StreamConn, error) {
+	select {
+	case acceptResponse, ok := <-sl.acceptCh:
+		if !ok {
+			return nil, net.ErrClosed
+		}
+		return acceptResponse.conn, acceptResponse.err
+	case <-sl.closeCh:
+		return nil, net.ErrClosed
+	}
+}
+
+func (sl *virtualStreamListener) Close() error {
+	close(sl.closeCh)
+	return sl.onCloseFunc()
+}
+
+func (sl *virtualStreamListener) Addr() net.Addr {
+	return sl.listener.Addr()
+}
+
+type virtualPacketConn struct {
+	net.PacketConn
+	onCloseFunc OnCloseFunc
+}
+
+func (spc *virtualPacketConn) Close() error {
+	return spc.onCloseFunc()
+}
+
+type listenAddr struct {
+	ln          Listener
+	acceptCh    chan acceptResponse
+	onCloseFunc OnCloseFunc
+}
+
+type canCreateStreamListener interface {
+	NewStreamListener(onCloseFunc OnCloseFunc) StreamListener
+}
+
+var _ canCreateStreamListener = (*listenAddr)(nil)
+
+// NewStreamListener creates a new [StreamListener].
+func (la *listenAddr) NewStreamListener(onCloseFunc OnCloseFunc) StreamListener {
+	if ln, ok := la.ln.(StreamListener); ok {
+		return &virtualStreamListener{
+			listener:    ln,
```

Comment:
All of these parameters are the same for all derived virtualListeners.
So there's no point in making different ones. You can keep returning the same!
Which means you can get rid of listenAddr, and use the virtual listener directly.